File: python/CombatRules.py
from Character import Character
from Checkers import AlliesChecker
from Checkers import DamageChecker
from Checkers import SelfAttackChecker

class CombatRules:
    def __init__(self, player_a, player_b: Character) -> None:
        self._player_a = player_a
        self._player_b = player_b
    
    def deal_damage(self, damage: int) -> None:
        try:
            checker = AlliesChecker(SelfAttackChecker(DamageChecker(None, damage)))
            checker.execute_check(self._player_a, self._player_a)
        except TypeError:
            print("AbstractChecker should be provided")

File: cpp/combat_rules.cpp
#include "combat_rules.hpp"
#include "abstract_checker.hpp"
#include "checkers.hpp"

CombatRules::CombatRules(const Character& playerA, const Character& playerB) : _playerA{playerA}, _playerB{playerB} {
}

CombatRules::CombatRules() {}

void CombatRules::dealDamage(int damage) {
    AbstractChecker* alliesChecker = new AlliesChecker();
    AbstractChecker* selfAttackChecker = new SelfAttackChecker();
    AbstractChecker* positionChecker = new PositionChecker();
    AbstractChecker* damageChecker = new DamageChecker(damage);

    // set all next
    alliesChecker->setNextChecker(selfAttackChecker);
    selfAttackChecker->setNextChecker(positionChecker);
    positionChecker->setNextChecker(damageChecker);

    // call the chain
    alliesChecker->performCheck(this->_playerA, this->_playerA);
}

void CombatRules::healDamage(int heal) {
}